src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.Split.cs

```diff
@@ -0,0 +1,152 @@
+                None = 0,
+                SingleToken,
+                Sequence,
+                Any,
+                SearchValues
+            }
+
+            private readonly ReadOnlySpan<T> _span;
+
+            private readonly T _separator = default!;
+            private readonly ReadOnlySpan<T> _separatorBuffer;
+            private readonly SearchValues<T> _searchValues = default!;
+
+            private readonly int _separatorLength;
+            private readonly SplitMode _splitMode;
+
+            private readonly bool _isInitialized = true;
+
+            private int _startCurrent = 0;
+            private int _endCurrent = 0;
+            private int _startNext = 0;
+
+            /// <summary>
+            /// Returns an enumerator that allows for iteration over the split span.
+            /// </summary>
+            /// <returns>Returns a <see cref="System.MemoryExtensions.SpanSplitEnumerator{T}"/> that can be used to iterate over the split span.</returns>
+            public SpanSplitEnumerator<T> GetEnumerator() => this;
+
+            /// <summary>
+            /// Returns the current element of the enumeration.
+            /// </summary>
+            /// <returns>Returns a <see cref="System.Range"/> instance that indicates the bounds of the current element withing the source span.</returns>
+            public Range Current => new Range(_startCurrent, _endCurrent);
+
+            internal SpanSplitEnumerator(ReadOnlySpan<T> span, SearchValues<T> searchValues)
+            {
+                _span = span;
+                _separatorLength = 1;
+                _splitMode = SplitMode.SearchValues;
+                _searchValues = searchValues;
+            }
+
+            internal SpanSplitEnumerator(ReadOnlySpan<T> span, ReadOnlySpan<T> separator, bool treatAsSingleSeparator)
+            {
+                _span = span;
+                _separatorBuffer = separator;
+                _separatorLength = (_separatorBuffer.Length, treatAsSingleSeparator) switch
+                {
+                    (0, true) or (_, false) => 1,
+                    _ => separator.Length
+                };
+                _splitMode = treatAsSingleSeparator ? SplitMode.Sequence : SplitMode.Any;
+            }
+
+            internal SpanSplitEnumerator(ReadOnlySpan<T> span, T separator)
+            {
+                _span = span;
+                _separator = separator;
+                _separatorLength = 1;
+                _splitMode = SplitMode.SingleToken;
+            }
+
+            /// <summary>
+            /// Advances the enumerator to the next element of the enumeration.
+            /// </summary>
+            /// <returns><see langword="true"/> if the enumerator was successfully advanced to the next element; <see langword="false"/> if the enumerator has passed the end of the enumeration.</returns>
+            public bool MoveNext()
+            {
+                if (!_isInitialized || _startNext > _span.Length)
```

Comment:
Why is _isInitialized needed?

Reply:
It's not inherently needed, but to ensure that the Enumerator returns `false` in the scenario of:
```csharp
SpanSplitEnumerator<char> e = default;
e.MoveNext(); // returns false
```
_some_ check is required, could of course also be `_splitMode != SplitMode.None`. I basically lifted-shifted the implementation from https://github.com/dotnet/runtime/pull/295 and I believe using `_isInitialized` was PR feedback from @layomia at the time.
Happy to align with whatever you feel is right.

Reply:
Even if we care about the unintialized enumerator case (we generally don't), you can bake that into the switch you already have for the mode, just having a mode of 0 be unintialized.